test_fms/coupler/test_coupler_types.F90

```diff
@@ -0,0 +1,311 @@
+array_2d = 1.0; array_3d = 1.0
+call check_field_data_2d(bc_2d_cp, array_2d)
+call check_field_data_3d(bc_3d_cp, array_3d)
+call coupler_type_destructor(bc_2d_cp)
+call coupler_type_destructor(bc_3d_cp)
+! 3d -> 2d, 3d
+call coupler_type_copy(bc_3d_new, bc_2d_cp, data_grid(1), data_grid(2), data_grid(3), data_grid(4), " ", &
+                       (/ NULL_AXIS_ID /), time_t )
+call coupler_type_copy(bc_3d_new, bc_3d_cp, data_grid(1), data_grid(2), data_grid(3), data_grid(4), data_grid(5), " ", &
+                       (/ NULL_AXIS_ID /), time_t )
+call coupler_type_copy_data(bc_3d_new, bc_3d_cp)
+call check_field_data_3d(bc_3d_cp, array_3d)
+call coupler_type_destructor(bc_2d_cp)
+call coupler_type_destructor(bc_3d_cp)
+
+! coupler_type_set_diags and coupler_type_send_data
+! set up for diag manager
+call diag_manager_init
+allocate(lats(1:nlat), lons(1:nlon), nzs(1:nz))
+do i=1, nlat
+  lats(i) = i
+enddo
+do i=1, nlon
+  lons(i) = i
+enddo
+do i=1, nz
+  nzs(i) = i
+enddo
+id_x  = diag_axis_init('x',  lats,  'point_E', 'x', long_name='point_E', Domain2=Domain)
+id_y  = diag_axis_init('y',  lons,  'point_N', 'y', long_name='point_N', Domain2=Domain)
+id_z  = diag_axis_init('z',  nzs,  'point_Z', 'z', long_name='point_Z')
+! registers field with data in type
+! reset the time and assign names to each field
+time_t = set_date(1, 1, 1)
+do i=1, num_bc
+  do j=1, num_fields
+    bc_2d_new%FMS_TEST_BC_TYPE_(i)%field(j)%name = "bc"//string(i)//"_var2d_"//string(j)
+    bc_3d_new%FMS_TEST_BC_TYPE_(i)%field(j)%name = "bc"//string(i)//"_var3d_"//string(j)
+    bc_2d_new%FMS_TEST_BC_TYPE_(i)%field(j)%long_name = "bc"//string(i)//"_variable_2d_"//string(j)//"_min"
+    bc_3d_new%FMS_TEST_BC_TYPE_(i)%field(j)%long_name = "bc"//string(i)//"_variable_3d_"//string(j)//"_min"
+  enddo
+enddo
+call coupler_type_set_diags(bc_2d_new, "test_coupler_types", (/id_x, id_y/), time_t)
+call coupler_type_set_diags(bc_3d_new, "test_coupler_types", (/id_x, id_y, id_z/), time_t)
+call coupler_type_copy(bc_2d_new, bc_2d_cp, data_grid(1), data_grid(2), data_grid(3), data_grid(4), " ", &
+                       (/null_axis_id/), time_t)
+call coupler_type_copy_data(bc_2d_new, bc_2d_cp)
+call coupler_type_copy(bc_3d_new, bc_3d_cp, data_grid(1), data_grid(2), data_grid(3), data_grid(4), data_grid(5),  " ",&
+                       (/null_axis_id/), time_t)
+call coupler_type_copy_data(bc_3d_new, bc_3d_cp)
+
+do day=1,31
+  time_t = set_date(1, 1, day)
+  call coupler_type_increment_data(bc_2d_cp, bc_2d_new) ! increment _new with cp
+  call coupler_type_increment_data(bc_3d_cp, bc_3d_new)
+  call coupler_type_send_data(bc_2d_new, time_t)
+  call coupler_type_send_data(bc_3d_new, time_t)
+enddo
+time_t = set_date(1, 2, 1)
+call diag_manager_end(time_t)
+
+! coupler_type_data_override
+! basic grid spec points to outputted .nc's
+if( mpp_pe() .eq. mpp_root_pe()) then
+  err = nf90_create('INPUT/grid_spec.nc', ior(nf90_clobber, nf90_64bit_offset), ncid)
```

Comment:
mainly because the data_override tests use netcdf directly and i was going off those files. i think its better this way actually even though it wasn't intentional, since unit test shouldn't be affected by modules they don't need to use.